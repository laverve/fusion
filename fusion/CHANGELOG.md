# Changelog

## [1.0.1](https://github.com/laverve/fusion/compare/pixi-fusion-v1.0.0...pixi-fusion-v1.0.1) (2024-10-14)


### Bug Fixes

* **fusion:** ignore errors during reinitialization ([1d663d3](https://github.com/laverve/fusion/commit/1d663d39add96dd3ce80f157d97597af310d7ce6))

## [1.0.0](https://github.com/laverve/fusion/compare/pixi-fusion-v1.0.0...pixi-fusion-v1.0.0) (2024-10-13)


### ⚠ BREAKING CHANGES

* creating a new release after cleaning up

### Features

* **pixi.js:** migrates to the lates version + restructuring repo ([b63d3be](https://github.com/laverve/fusion/commit/b63d3bee1a53ee7933b67b8e0574701b051b9186))


### Bug Fixes

* **fusion:** dummy commit to initiate normal release for fusion ([47fc295](https://github.com/laverve/fusion/commit/47fc2954f9a6a6511b92a645bbf7cd53279182ca))


### Miscellaneous Chores

* bump version ([5000aab](https://github.com/laverve/fusion/commit/5000aaba0487d91b51c023333dd07637167cc221))
* release 0.0.10 ([bd91141](https://github.com/laverve/fusion/commit/bd91141158bc1b90cd36315691163b22c681816b))
* release 1.0.0 ([b653a83](https://github.com/laverve/fusion/commit/b653a830d642f1694ab98eca867f5a82c8951eb8))
* release 3.0.0 ([bcb0164](https://github.com/laverve/fusion/commit/bcb0164e2024fab9bca5f217dc54ecea8f6ca4e5))

## [1.0.0](https://github.com/laverve/fusion/compare/fusion-v0.0.10...fusion-v1.0.0) (2024-10-13)


### ⚠ BREAKING CHANGES

* creating a new release after cleaning up

### Features

* **pixi.js:** migrates to the lates version + restructuring repo ([b63d3be](https://github.com/laverve/fusion/commit/b63d3bee1a53ee7933b67b8e0574701b051b9186))


### Bug Fixes

* **fusion:** dummy commit to initiate normal release for fusion ([47fc295](https://github.com/laverve/fusion/commit/47fc2954f9a6a6511b92a645bbf7cd53279182ca))


### Miscellaneous Chores

* bump version ([5000aab](https://github.com/laverve/fusion/commit/5000aaba0487d91b51c023333dd07637167cc221))
* release 0.0.10 ([bd91141](https://github.com/laverve/fusion/commit/bd91141158bc1b90cd36315691163b22c681816b))
* release 1.0.0 ([b653a83](https://github.com/laverve/fusion/commit/b653a830d642f1694ab98eca867f5a82c8951eb8))
* release 3.0.0 ([bcb0164](https://github.com/laverve/fusion/commit/bcb0164e2024fab9bca5f217dc54ecea8f6ca4e5))

## [0.0.10](https://github.com/laverve/fusion/compare/fusion-v3.0.1...fusion-v0.0.10) (2024-10-13)


### Miscellaneous Chores

* release 0.0.10 ([bd91141](https://github.com/laverve/fusion/commit/bd91141158bc1b90cd36315691163b22c681816b))

## [3.0.1](https://github.com/laverve/fusion/compare/fusion-v3.0.0...fusion-v3.0.1) (2024-10-13)


### Bug Fixes

* **fusion:** dummy commit to initiate normal release for fusion ([47fc295](https://github.com/laverve/fusion/commit/47fc2954f9a6a6511b92a645bbf7cd53279182ca))
