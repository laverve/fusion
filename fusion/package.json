{
    "author": "Victor Pishuk <victor.pishuk@gmail.com>",
    "type": "module",
    "bugs": {
        "url": "https://github.com/laverve/fusion/issues"
    },
    "dependencies": {
        "@types/matter-js": "^0.19.7",
        "matter-js": "^0.20.0",
        "pixi.js": "^8.4.1"
    },
    "peerDependencies": {
        "react": "^18.2.0"
    },
    "description": "This module offers a set of common components needed for playing games.",
    "keywords": [
        "game",
        "engine",
        "pixi.js",
        "matter-js",
        "react"
    ],
    "license": "MIT",
    "main": "./dist/src/index.js",
    "types": "./dist/src/index.d.ts",
    "name": "@laverve/fusion",
    "files": [
        "./dist/src/",
        "./src/",
        "package.json",
        "package-lock.json",
        "README.md",
        "LICENSE"
    ],
    "publishConfig": {
        "access": "public"
    },
    "repository": {
        "type": "git",
        "url": "https://github.com/laverve/fusion.git"
    },
    "scripts": {
        "lint": "eslint .",
        "lint:staged": "lint-staged",
        "test": "jest --passWithNoTests",
        "build": "rimraf ./dist && tsc"
    },
    "version": "3.0.0",
    "webpack": "./src/index.ts"
}
