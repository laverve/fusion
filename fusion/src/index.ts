export * from "./game-context";
export * from "./assets-manager";
export * from "./world";
export * from "./camera";
export * from "./layer";
export * from "./assets-manager";
export * from "./game-objects";
export * from "./physics";
export * from "./types";
export * from "./hooks";
