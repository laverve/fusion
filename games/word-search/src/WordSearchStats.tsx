import React, { useContext } from "react";
import classnames from "classnames";
import { useTranslation } from "react-i18next";

import { useGame } from "pixi-fusion";
import { WordSearchContext } from "./WordSearch.context";

export type WordSearchStatsProps = { classNames?: string };

export const WordSearchStats: React.FC<WordSearchStatsProps> = ({ classNames = "" }) => {
    const { foundWords } = useContext(WordSearchContext);
    const { t } = useTranslation();

    const { startTime, endTime } = useGame();

    const completedInSeconds = Math.round(((endTime || 0) - (startTime || 0)) / 1000);
    const seconds = completedInSeconds % 60;
    const minutes = Math.floor(completedInSeconds / 60);

    return (
        <div className={classnames("flex", "flex-col", classNames)} data-testid="testid-wordsearch-stats">
            <div className="flex flex-row justify-stretch gap-1">
                <p className="body-2 flex-1 text-right">
                    <strong>{t("controls.stats.foundWords")}</strong>
                </p>
                <p className="body-2 flex-1 text-center">{foundWords.length}</p>
            </div>
            {startTime && endTime && (
                <div className="flex flex-row justify-stretch gap-1">
                    <p className="body-2 flex-1 text-right">
                        {" "}
                        <strong>{t("controls.stats.timeSpent")}</strong>
                    </p>
                    <p className="body-2 flex-1 text-center">
                        {minutes < 10 && "0"}
                        {minutes} {t("controls.stats.minutes")} {seconds < 10 && "0"}
                        {seconds} {t("controls.stats.seconds")}
                    </p>
                </div>
            )}
        </div>
    );
};
