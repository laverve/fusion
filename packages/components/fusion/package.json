{
    "author": "Victor Pishuk <victor.pishuk@gmail.com>",
    "bugs": {
        "url": "https://github.com/laverve/fusion/issues"
    },
    "dependencies": {
        "@types/matter-js": "^0.19.2",
        "matter-js": "^0.19.0",
        "pixi.js": "^7.3.2"
    },
    "peerDependencies": {
        "react": "^18.2.0"
    },
    "description": "This module offers a set of common components needed for playing games.",
    "keywords": [
        "game",
        "engine",
        "pixi.js",
        "matter-js",
        "react"
    ],
    "license": "MIT",
    "main": "./dist/src/index.js",
    "types": "./dist/src/index.d.ts",
    "name": "@laverve/fusion",
    "files": [
        "./dist/src/",
        "./src/",
        "package.json",
        "package-lock.json",
        "README.md",
        "LICENSE"
    ],
    "publishConfig": {
        "access": "public"
    },
    "repository": {
        "type": "git",
        "url": "https://github.com/laverve/fusion.git"
    },
    "scripts": {
        "lint": "eslint --ext .js,.ts,.tsx,.json .",
        "lint:staged": "lint-staged",
        "test": "jest --passWithNoTests",
        "build": "tsc"
    },
    "version": "2.0.2",
    "webpack": "./src/index.ts"
}
