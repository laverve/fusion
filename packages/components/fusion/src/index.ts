export * from "./game-context";
export * from "./physics";
export * from "./stage";
export * from "./game-objects";
export * from "./types";
