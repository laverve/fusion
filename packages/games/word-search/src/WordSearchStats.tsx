import React, { useContext } from "react";
import classnames from "classnames";
import { useTranslation } from "react-i18next";
import { Typography } from "@laverve/design-provider";

import { GameContext } from "@laverve/fusion";
import { WordSearchContext } from "./WordSearch.context";
import { I18N_NAMESPACE } from "./types";

export type WordSearchStatsProps = { classNames?: string };

export const WordSearchStats: React.FC<WordSearchStatsProps> = ({ classNames = "" }) => {
    const { foundWords } = useContext(WordSearchContext);
    const { t } = useTranslation(I18N_NAMESPACE);

    const { startTime, endTime } = useContext(GameContext);

    const completedInSeconds = Math.round(((endTime || 0) - (startTime || 0)) / 1000);
    const seconds = completedInSeconds % 60;
    const minutes = Math.floor(completedInSeconds / 60);

    return (
        <div className={classnames("flex", "flex-col", classNames)} data-testid="testid-wordsearch-stats">
            <div className="flex flex-row justify-stretch gap-1">
                <Typography variant="body-2" className="flex-1 text-right">
                    <strong>{t("controls.stats.foundWords")}</strong>
                </Typography>
                <Typography variant="body-2" className="flex-1 text-center">
                    {foundWords.length}
                </Typography>
            </div>
            {startTime && endTime && (
                <div className="flex flex-row justify-stretch gap-1">
                    <Typography variant="body-2" className="flex-1 text-right">
                        {" "}
                        <strong>{t("controls.stats.timeSpent")}</strong>
                    </Typography>
                    <Typography variant="body-2" className="flex-1 text-center">
                        {minutes < 10 && "0"}
                        {minutes} {t("controls.stats.minutes")} {seconds < 10 && "0"}
                        {seconds} {t("controls.stats.seconds")}
                    </Typography>
                </div>
            )}
        </div>
    );
};
