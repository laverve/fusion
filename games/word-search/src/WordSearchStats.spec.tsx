/**
 * @jest-environment jsdom
 */

import { describe, it, expect } from "@jest/globals";
import React, { useContext, useState } from "react";

import { render, fireEvent } from "@testing-library/react";
import { GameContextProvider, GameContext } from "@laverve/fusion";

import { WordSearchContextProvider, WordSearchContext } from "./WordSearch.context";
import { WordSearchStats, WordSearchStatsProps } from "./WordSearchStats";

export const TestGameplay: React.FC<{ words: string[] }> = ({ words }) => {
    const { onWordFound } = useContext(WordSearchContext);
    const { start, stop } = useContext(GameContext);
    const [idx, setIdx] = useState(0);
    return (
        <>
            <button
                type="button"
                data-testid="test-on-word-found"
                onClick={() => {
                    onWordFound(words[idx]);
                    setIdx(idx + 1);
                }}
            >
                Find word
            </button>
            <button
                type="button"
                data-testid="test-start"
                onClick={() => {
                    start();
                }}
            >
                Start
            </button>
            <button
                type="button"
                data-testid="test-stop"
                onClick={() => {
                    stop();
                }}
            >
                Stop
            </button>
        </>
    );
};

describe("WordSearchStats", () => {
    const WORDS = ["dog", "cat", "dig", "boat", "doc"];
    const GRID = [
        ["b", "d", "i", "g"],
        ["d", "o", "g", "o"],
        ["u", "c", "a", "t"],
        ["x", "y", "z", "t"]
    ];

    const renderComponentUderTest = ({
        props = {},
        timeout = 0,
        selectedWords = ["cat"],
        selectedWordsColors = ["black"]
    }: {
        props?: WordSearchStatsProps;
        timeout?: number;
        selectedWords?: string[];
        selectedWordsColors?: string[];
    } = {}) => {
        return render(
            <GameContextProvider timeout={timeout}>
                <WordSearchContextProvider grid={GRID} words={WORDS} selectedWordsColors={selectedWordsColors}>
                    <TestGameplay words={selectedWords} />
                    <WordSearchStats {...props} />
                </WordSearchContextProvider>
            </GameContextProvider>
        );
    };

    it("should render only found words if game has not been started or stopped yet", () => {
        const component = renderComponentUderTest({
            selectedWords: WORDS
        });
        WORDS.forEach(() => {
            fireEvent.click(component.getByTestId("test-on-word-found"));
        });
        expect(component.asFragment()).toMatchSnapshot();
    });

    it("should render found words and time spent after game was started and stopped", () => {
        const component = renderComponentUderTest({
            selectedWords: WORDS
        });
        fireEvent.click(component.getByTestId("test-start"));
        WORDS.forEach(() => {
            fireEvent.click(component.getByTestId("test-on-word-found"));
        });
        fireEvent.click(component.getByTestId("test-stop"));
        expect(component.asFragment()).toMatchSnapshot();
    });
});
