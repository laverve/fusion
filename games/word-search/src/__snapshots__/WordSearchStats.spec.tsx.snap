// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordSearchStats should render found words and time spent after game was started and stopped 1`] = `
<DocumentFragment>
  <div
    data-overlay-container="true"
  >
    <button
      data-testid="test-on-word-found"
      type="button"
    >
      Find word
    </button>
    <button
      data-testid="test-start"
      type="button"
    >
      Start
    </button>
    <button
      data-testid="test-stop"
      type="button"
    >
      Stop
    </button>
    <div
      class="flex flex-col"
      data-testid="testid-wordsearch-stats"
    >
      <div
        class="flex flex-row justify-stretch gap-1"
      >
        <p
          class="body-2 flex-1 text-right"
        >
          <strong>
            Found words:
          </strong>
        </p>
        <p
          class="body-2 flex-1 text-center"
        >
          5
        </p>
      </div>
      <div
        class="flex flex-row justify-stretch gap-1"
      >
        <p
          class="body-2 flex-1 text-right"
        >
           
          <strong>
            Time spent:
          </strong>
        </p>
        <p
          class="body-2 flex-1 text-center"
        >
          00 m. 00 s.
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`WordSearchStats should render only found words if game has not been started or stopped yet 1`] = `
<DocumentFragment>
  <div
    data-overlay-container="true"
  >
    <button
      data-testid="test-on-word-found"
      type="button"
    >
      Find word
    </button>
    <button
      data-testid="test-start"
      type="button"
    >
      Start
    </button>
    <button
      data-testid="test-stop"
      type="button"
    >
      Stop
    </button>
    <div
      class="flex flex-col"
      data-testid="testid-wordsearch-stats"
    >
      <div
        class="flex flex-row justify-stretch gap-1"
      >
        <p
          class="body-2 flex-1 text-right"
        >
          <strong>
            Found words:
          </strong>
        </p>
        <p
          class="body-2 flex-1 text-center"
        >
          5
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
